src/what-unsafe-does.md

```diff
@@ -54,18 +53,97 @@ empty, "dangling" is the same as "non-null". Note that slices point to their
 entire range, so it's very important that the length metadata is never too
 large. If for some reason this is too cumbersome, consider using raw pointers.
 
-That's it. That's all the causes of Undefined Behavior baked into Rust. Of
-course, unsafe functions and traits are free to declare arbitrary other
-constraints that a program must maintain to avoid Undefined Behavior. For
-instance, the allocator APIs declare that deallocating unallocated memory is
-Undefined Behavior.
 
-However, violations of these constraints generally will just transitively lead to one of
-the above problems. Some additional constraints may also derive from compiler
-intrinsics that make special assumptions about how code can be optimized. For instance,
-Vec and Box make use of intrinsics that require their pointers to be non-null at all times.
 
-Rust is otherwise quite permissive with respect to other dubious operations.
+## Invalid Values: Yes We Mean It
+
+Many have trouble accepting the consequences of invalid values, so they merit
+some extra discussion here so no one misses it. The claim being made here is a
+very strong and surprising one, so read carefully.
+
+A value is *produced* whenever it is assigned, passed to something, or returned
+from something. Keep in mind references get to assume their referents are valid,
+so you can't even create a reference to an invalid value.
+
+Additionally, [uninitialized memory][] is **always invalid**, so you can't assign it to
```

Comment:
See above, this is incorrect in its generality. You are restricting this later, but I think that's too late. Also, first saying one thing and then later "well but we didn't really mean" it is really confusing.

Reply:
This is a fair criticism. I wanted to try this approach out, but it's definitely worth rethinking.

Basically I kinda like this approach because, again, our primary interest is in preventing programmers from doing bad things. So "you can't do this" followed by "...except for here" isn't a *terrible* approach with that goal. 

If people bounce off, they come away with a hyper-conservative model and try to avoid messing with uninitialized memory, which is good imo.

If people see that and get confused/angry, they can keep reading and go "aha!".

Reply:
In my experience (which is mostly based on writing scientific papers), one has to be *very careful* when leading with a wrong statement and correcting that later. This is definitely sometimes a good strategy, but it's a double-edged sword when you also consider e.g. your reader's faith in you. Usually, the least I'd do in a case like this is to add a footnote (would have to be a parenthetical here because footnotes are rendered too far away) saying something like "this is not strictly correct, we will refine this statement later". This at least prepares the reader for the blow coming later.

That said, in this case, I disagree with the entire approach. I think it is wrong to call out uninitialized memory as anything special. The fact that an uninitialized `bool` is UB should fall out of the same general principle that makes a NULL reference UB. We can have the desired rules for uninitialized memory based just on the notion of "valid value", with no need for programmers to learn *another* not-entirely-but-somewhat orthogonal concept.

On top of that, we should mentally prepare for when https://github.com/rust-lang/unsafe-code-guidelines/issues/71 gets resolved. The likely resolution (and IMO the best one) is that we will declare uninitialized integers as *not being UB*. Only when an uninitialized integer is fed into a primitive operation (like `+` or `|` or `==`), *then* we have UB if there is anything uninitialized. With that in mind, I think the route you are trying here is even less appealing.

Reply:
I was under the impression we were *very* far from resolving the uninit integer question satisfactorily. If that's not the case, I agree we should focus on a tighter definition.

Reply:
I won't dare making predictions. But the person arguing most strongly *against* allowing uninit integers was me, and I changed my mind. gnzlbg went back and forth, not sure what their current stanza is. I don't actually know any argument *for* making uninit integers insta-UB, besides the few that I opened the thread with, and I don't consider them convincing enough (any more) given the benefits of allowing uninit integers (mostly, everyone does it anyway^^).

If I had infinite time, the RFC would already have been written. I consider this the least controversial amongst the open questions around validity that we have. I might be missing a controversy though.